File: mirror/tier2-ruby/showcase/async_coroutines.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Async & Control Flow
# Module: async_coroutines.rb
# Purpose: Demonstrate coroutine patterns

class Coroutine
  def initialize(&block)
    @fiber = Fiber.new(&block)
    @state = "suspended"
  end
  
  def resume(value = nil)
    @state = "running"
    begin
      result = @fiber.resume(value)
      @state = "suspended"
      result
    rescue FiberError
      @state = "dead"
      nil
    end
  end
  
  def get_state
    @state
  end
end

def simple_coroutine
  value = Fiber.yield
  Fiber.yield value * 10
end

result = {
  coroutine_state_init: "suspended",
  coroutine_active: true,
}

puts result.inspect

File: mirror/tier2-python/showcase/async_coroutines.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Async & Control Flow
Module: async_coroutines.py
Purpose: Demonstrate coroutine patterns
"""

from typing import Generator, Any


class Coroutine:
    """Coroutine-like class"""
    def __init__(self, generator: Generator):
        self.gen = generator
        self.state = "suspended"
    
    def resume(self, value: Any = None):
        """Resume coroutine"""
        self.state = "running"
        try:
            result = self.gen.send(value)
            self.state = "suspended"
            return result
        except StopIteration:
            self.state = "dead"
            return None
    
    def get_state(self):
        """Get current state"""
        return self.state


def coroutine_generator():
    """Simple generator coroutine"""
    x = yield
    y = yield x * 2
    z = yield x + y
    yield z * 2


def simple_coroutine():
    """Simpler coroutine"""
    value = yield
    yield value * 10


result = {
    "coroutine_state_init": "suspended",
    "coroutine_active": True,
}

if __name__ == "__main__":
    print(result)
